odl/space/pspace.py

```diff
@@ -650,11 +650,41 @@ def __getitem__(self, indices):
 
     def __setitem__(self, indices, values):
         """Implement ``self[indices] = values``."""
+        # Get the parts to which we assign values
+        if isinstance(indices, Integral):
+            indexed_parts = (self.parts[indices],)
+            values = (values,)
+        elif isinstance(indices, slice):
+            indexed_parts = self.parts[indices]
+        elif isinstance(indices, list):
+            indexed_parts = tuple(self.parts[i] for i in indices)
+        else:
+            raise TypeError('bad index type {}'.format(type(indices)))
+
+        # Do the assignment, with broadcasting if desired
         try:
-            self.parts[indices] = values
+            iter(values)
         except TypeError:
-            for i, index in enumerate(indices):
-                self.parts[index] = values[i]
+            # `values` is not iterable, assume it can be assigned to
+            # all indexed parts
+            for p in indexed_parts:
+                p[:] = values
+        else:
+            # `values` is iterable; it could still represent a single
+            # element of a power space.
+            if self.space.is_power_space and values in self.space[0]:
```

Comment:
What about stuff like assigning a list? I think that works in rn, i.e.

```
>>> r2 = odl.rn(2)
>>> el = r2.element()
>>> el[:] = [1, 2]
```

don't we want something like that to work here?